File: ant_books/2_3/dp2.rb
N = 4
W = [2, 1, 3, 2]
V = [3, 2, 4, 2]
WEIGHT = 5

DP = Array.new(N+1).map { Array.new(WEIGHT+1, 0) }

def solve
  (N-1).step(0, -1) do |i|
    for j in 0..WEIGHT
      if j < W[i]
        DP[i][j] = DP[i + 1][j]
      else
        DP[i][j] = [DP[i + 1][j], DP[i+1][j-W[i]] + V[i]].max
      end
    end
  end
  puts DP[0][WEIGHT];
end

solve

File: ant_books/2_3/dp2.rs
const N: usize = 4;
const WS: [usize; 4] = [2, 1, 3, 2];
const VS: [usize; 4] = [3, 2, 4, 2];
const W: usize = 5;

use std::cmp;

fn solve() -> usize {
  let mut dp = vec![ vec![0; W+1]; N+1];

  for i in (0..N).rev() {
    for j in 0..W+1 {
      if j < WS[i] {
        dp[i][j] = dp[i + 1][j];
      } else {
        dp[i][j] = cmp::max(dp[i + 1][j], dp[i+1][j-WS[i]] + VS[i]);
      }
    }
  }
  dp[0][W]
}

fn main() {
  println!("{}", solve());
}